contracts/core/Stake.sol

```diff
@@ -341,41 +427,107 @@ contract Stake is StakeInterface {
         validatorWeight_ = validatorWeightAtHeight(_height, _validator);
     }
 
+    /**
+     * @notice Returns all registered validator addresses.
+     *
+     * @return An array of validator addresses. Includes evicted validators.
+     */
+    function validatorAddresses() external view returns (address[]) {
+        return validatorAddresses;
+    }
+
     /* Private Functions */
 
+    /**
+     * @notice Calculates the total weight of all active validators at a given
+     *         height.
+     *
+     * @param _height For which height to get the total weight.
+     *
+     * @return totalWeight_ At the given height.
+     */
+    function totalWeightAtHeight_(
+        uint256 _height
+    )
+        private
+        view
+        returns (uint256 totalWeight_)
+    {
+        for (uint256 i = 0; i < validatorAddresses.length; i++) {
+            address validator = validatorAddresses[i];
+            totalWeight_ += validatorWeightAtHeight(_height, validator);
+        }
```

Comment:
This can potentially create out of gas error.
`totalWeightAtHeight_` is looping throught all the validators, we should store the total weight for each height or for some checkpoint heights so that the iteration will not result in out of gas.

We may consider optimization in future.

Reply:
Yes you are right. This is addressed in #395